### PR TITLE
util: resurrect UdpFramed

### PR DESCRIPTION
**Pull request by leshow**
## Motivation
Brings back `UdpFramed` in `tokio-util`. This was the only way to create a `Stream`/`Sink` from a `UdpSocket` using `Decoder`/`Encoder` so I think it's important to have.

In earlier commits I had also brought back `split` for `UdpSocket` but I've removed that for now. I wanted a way to `split` on `UdpFramed` also and get an independent `Stream`/`Sink`, but there's no way to share the `Codec` among both sides that I can see, since they both require `&mut` access. So for now, just bringing back `UdpFramed`

## Solution

Mostly just un-commenting what was there with some small modifications due to `ReadBuf`

I think this closes out #2830 , we've added back `poll_*` functions for `UdpSocket` and `UdpFramed`

### CODE REVIEW

File: tokio-util/Cargo.toml [hunk 1]
```diff
@@ -24,8 +24,9 @@ categories = ["asynchronous"]
 default = []
 
 # Shorthand for enabling everything
-full = ["codec", "compat", "io", "time"]
+full = ["codec", "compat", "io", "time", "udp"]
 
+udp = ["tokio/net"]
```
Comment (taiki-e) [inline tokio-util/Cargo.toml:29]: Given that `tokio` crate merged `udp` feature into `net` feature, this may be preferable to be a `net` feature.

Comment (leshow) [reply tokio-util/Cargo.toml:29]: Certainly, I went back and forth a few times on the naming myself. I'm happy to change to `net`

File: tokio-util/src/cfg.rs [hunk 2]
```diff
@@ -18,17 +18,15 @@ macro_rules! cfg_compat {
...
 
-/*
-macro_rules! cfg_udp {
+macro_rules! cfg_net {
     ($($item:item)*) => {
         $(
-            #[cfg(all(feature = "udp", feature = "codec"))]
-            #[cfg_attr(docsrs, doc(cfg(all(feature = "udp", feature = "codec"))))]
+            #[cfg(all(feature = "net", feature = "codec"))]
+            #[cfg_attr(docsrs, doc(cfg(all(feature = "net", feature = "codec"))))]
```
Comment (Darksonn) [inline tokio-util/src/cfg.rs:25]: It seems a bit weird that `cfg_net!` is matched under `net + codec`.

File: tokio-util/src/udp/frame.rs [hunk 3]
```diff
@@ -69,13 +71,12 @@ impl<C: Decoder + Unpin> Stream for UdpFramed<C> {
             let addr = unsafe {
                 // Convert `&mut [MaybeUnit<u8>]` to `&mut [u8]` because we will be
                 // writing to it via `poll_recv_from` and therefore initializing the memory.
-                let buf: &mut [u8] =
-                    &mut *(pin.rd.bytes_mut() as *mut [MaybeUninit<u8>] as *mut [u8]);
-
-                let res = ready!(Pin::new(&mut pin.socket).poll_recv_from(cx, buf));
+                let buf: &mut [u8] = &mut *(pin.rd.bytes_mut() as *mut _ as *mut [u8]);
+                let mut read = ReadBuf::new(buf);
+                let res = ready!(Pin::new(&mut pin.socket).poll_recv_from(cx, &mut read));
```
Comment (Darksonn) [inline tokio-util/src/udp/frame.rs:76]: This should use `ReadBuf::uninit`.

File: tokio-util/src/cfg.rs (hunk 2 above)
Comment (leshow) [reply tokio-util/src/cfg.rs:25]: It was called `cfg_udp` before, I was just sticking with that naming. Does tokio have other `cfg!` macros that have multiple features we can look to for consistent naming?

File: tokio-util/src/udp/frame.rs (hunk 3 above)
Comment (leshow) [reply tokio-util/src/udp/frame.rs]: Good catch, thank you!

File: tokio-util/src/udp/frame.rs [hunk 4]
```diff
@@ -69,13 +71,12 @@ impl<C: Decoder + Unpin> Stream for UdpFramed<C> {
             let addr = unsafe {
                 // Convert `&mut [MaybeUnit<u8>]` to `&mut [u8]` because we will be
                 // writing to it via `poll_recv_from` and therefore initializing the memory.
-                let buf: &mut [u8] =
-                    &mut *(pin.rd.bytes_mut() as *mut [MaybeUninit<u8>] as *mut [u8]);
-
-                let res = ready!(Pin::new(&mut pin.socket).poll_recv_from(cx, buf));
+                let buf = &mut *(pin.rd.bytes_mut() as *mut _ as *mut [MaybeUninit<u8>]);
+                let mut read = ReadBuf::uninit(buf);
+                let res = ready!(Pin::new(&mut pin.socket).poll_recv_from(cx, &mut read));
```
Comment (Darksonn) [inline tokio-util/src/udp/frame.rs:76]: We should either assert that the `ReadBuf` is not swapped, or at least add a comment that we rely on the main Tokio crate doing the right thing here.

**Review by Darksonn — APPROVED** [review]

File: tokio-util/src/udp/frame.rs (hunk 4 above)
Comment (leshow) [reply tokio-util/src/udp/frame.rs:81]: Let's go with the assertion

**Pull request merged**